File: Solutions/Python/rotating_the_box.py
"""
    [MEDIUM]
    1861. Rotating the Box

    Concepts:
    - matrix
    - two pointers

    Stats:
        Runtime | 1917 ms   [Beats 45.17%]
        Memory  | 29.21 MB  [Beats 9.52%]
"""

class Solution:
    def rotateTheBox(self, box: List[List[str]]) -> List[List[str]]:
        """
            1) rotate matrix 90 degrees clockwise
            2) keep pointer of lowest empty space and pointer
                searching for stones
            3) when stone found, swap items at bottom
                and search pointers
        """
        m = len(box)
        n = len(box[0])

        # n rows, m cols
        res = [[0]*m for _ in range(n)]

        for i in range(m):
            for j in range(n):
                res[j][m-1-i] = box[i][j]

        # iterate over all columns
        for j in range(m):
            bottom = n - 1

            # iterate backwards up rows
            i = n - 1
            while i >= 0:
                # if at top of col, break
                if i < 0:
                    break

                # find lowest empty space for bottom
                if res[i][j] == "*":
                    bottom = i - 1

                # swap stone w lowest space and update
                elif res[i][j] == "#":
                    res[i][j], res[bottom][j] = res[bottom][j], res[i][j]
                    bottom -= 1

                i -= 1

        return res
                

File: Solutions/Rust/rotating_the_box.rs
/*
    [MEDIUM]
    1861. Rotating the Box

    Concepts:
    - matrix
    - two pointers

    Stats:
        Runtime | 10 ms     [Beats 57.14%]
        Memory  | 18.90 MB  [Beats 57.14%]
*/
impl Solution {
    pub fn rotate_the_box(bocks: Vec<Vec<char>>) -> Vec<Vec<char>> {
        let [m, n] = [bocks.len(), bocks[0].len()];
        let mut res = vec![vec!['.'; m]; n];

        for i in 0..m {
            for j in 0..n {
                res[j][m-1-i] = bocks[i][j];
            }
        }

        // iterate over all columns
        for j in 0..m {
            let [mut i, mut bottom] = [n - 1; 2];
            while i >= 0 {
                // find lowest empty space for bottom
                if res[i][j] == '*' { bottom = i - 1; }
                // swap stone w lowest space and update
                else if res[i][j] == '#' {
                    [res[i][j], res[bottom][j]] = [res[bottom][j], res[i][j]];
                    bottom -= 1;
                }

                // if at top of col, break
                if i == 0 { break; }
                i -= 1;
            }
        }

        return res;
    }
}